### PR TITLE
Add registry refreshing and caching

### PR DESCRIPTION
**Pull request by woop**
**What this PR does / why we need it**:
* Adds support for caching of the RegistryProto as part of the Registry. Used during `get_online_features()` to prevent users from having to download the registry on every call.
* Also adds a `refresh_registry()` method to the FeatureStore class for users that want to explicitly update their registries.

**Does this PR introduce a user-facing change?**:

```release-note
Added support for registry caching and added a refresh_registry() method to the FeatureStore class
```

### CODE REVIEW

**Comment by woop** [comment]
Blocked by #1432

File: sdk/python/feast/feature_store.py [hunk 1]
```diff
@@ -72,15 +73,31 @@ def __init__(
...
     def _get_provider(self) -> Provider:
         return get_provider(self.config)
 
-    def _get_registry(self) -> Registry:
-        return Registry(self.config.metadata_store)
+    def refresh_registry(self):
+        """Fetches and caches a copy of the feature registry in memory.
+
+        Explicitly calling this method makes it possible for methods like get_online_features to use cached registry
+        state instead of retrieving the registry state at request-time, thus reducing latency."""
```
Comment (jklegar) [inline sdk/python/feast/feature_store.py:93]: can we clarify a little more how this works, i.e. if this function is never called then every cache_ttl seconds the get_online_features request will refresh the cache but if this function is called it refreshes the cache and resets the timer back to cache_ttl seconds

File: sdk/python/feast/registry.py [hunk 2]
```diff
@@ -333,16 +372,22 @@ def __init__(self, filepath: str):
...
-        else:
+            return registry_proto
+        raise FileNotFoundError(
+            f'Registry not found at path "{self._filepath}". Have you run "feast apply"?'
+        )
+
+    def update_registry_proto(self, updater: Callable[[RegistryProto], RegistryProto]):
+        try:
+            registry_proto = self.get_registry_proto()
+        except FileNotFoundError:
```
Comment (jklegar) [inline sdk/python/feast/registry.py:387]: nice

**Review by jklegar — APPROVED** [review]
LGTM with one comment

**Pull request merged**